lib/command/src/Obelisk/Command/Preprocessor.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Obelisk.Command.Preprocessor where
+
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.ByteString.Lazy.Builder as BU
+import Data.List
+import Data.Maybe
+import qualified Data.Text.Lazy as TL
+import qualified Data.Text.Lazy.Builder as TL
+import qualified Data.Text.Lazy.Encoding as TL
+import Distribution.Compiler
+import Language.Haskell.Extension
+import System.Environment
+import System.IO
+import System.FilePath
+
+import Obelisk.App
+import Obelisk.Command (mkObeliskConfig)
+import Obelisk.Command.Run
+
+main :: IO ()
+main = do
+  -- The command line argumens are passed in via ghci.
+  -- ghci dictates the first three options and meanings
+  -- TODO: perform some validation and print a nice error/help message, if we don't see something plausible.
+  -- We do have control over the remaining arguments, but they must be the same for all files.
+  -- We assume all the remaining arguments passed in are paths to cabal or hpack package specifications.
+
+  args@(origPath:inPath:outPath:packagePaths) <- getArgs
+
+  outFile <- openFile outPath WriteMode
+  let hPutTextBuilder h = BU.hPutBuilder h . TL.encodeUtf8Builder . TL.toLazyText
+
+  putStr "--------------------------------------------------------------------------------\n"
+  print args
+  putStr "--------------------------------------------------------------------------------\n"
+
+  -- Thus we must select among the packagePaths for the file we are going to parse.
+
+  let takeDirs = takeWhile hasTrailingPathSeparator
+      packageDirs = sortOn (negate . length . takeDirs) $ map (splitPath . normalise) packagePaths
+      origDir = splitPath $ normalise $ origPath
+      matches = [ joinPath d | d <- packageDirs, takeDirs d `isPrefixOf` origDir ]
+
+  -- So the first element of matches is going to be the deepest path to a package spec that contains
+  -- our file as a subdirectory.
+
+  config <- mkObeliskConfig
+
+  case matches of
+    [] -> hPutTextBuilder outFile (lineNumberPragma origPath) -- TODO: probably should produce a warning
+    (packagePath:_) -> do
+       runObelisk config (parseCabalPackage packagePath) >>= \case
```

Comment:
Does it makes sense to pull `parseCabalPackage` out of `MonadObelisk` to avoid this dance?

Reply:
I would say, probably;  I haven't explored exactly why `parseCabalPackage` is in `MonadObelisk`, but I don't think there's anything important there.

Reply:
Yeah probably just for error throwing stuff.